sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java

```diff
@@ -0,0 +1,382 @@
+          value instanceof Long, "Expecting a value as Long type (millis).");
+      return (Long) value;
+    }
+  }
+
+  static Integer convertDate(Object value) {
+    if (value instanceof ReadableInstant) {
+      return Days.daysBetween(Instant.EPOCH, (ReadableInstant) value).getDays();
+    } else {
+      Preconditions.checkArgument(
+          value instanceof Integer, "Expecting a value as Integer type (days).");
+      return (Integer) value;
+    }
+  }
+
+  static ByteString convertDecimal(LogicalType logicalType, Object value) {
+    ByteBuffer byteBuffer = (ByteBuffer) value;
+    BigDecimal bigDecimal =
+        new Conversions.DecimalConversion()
+            .fromBytes(
+                byteBuffer.duplicate(),
+                Schema.create(Schema.Type.NULL), // dummy schema, not used
+                logicalType);
+    return BeamRowToStorageApiProto.serializeBigDecimalToNumeric(bigDecimal);
+  }
+
+  /**
+   * Given an Avro Schema, returns a protocol-buffer TableSchema that can be used to write data
+   * through BigQuery Storage API.
+   *
+   * @param schema An Avro Schema
+   * @return Returns the TableSchema created from the provided Schema
+   */
+  public static TableSchema protoTableSchemaFromAvroSchema(Schema schema) {
+    Preconditions.checkState(!schema.getFields().isEmpty());
+
+    TableSchema.Builder builder = TableSchema.newBuilder();
+    for (Schema.Field field : schema.getFields()) {
+      builder.addFields(fieldDescriptorFromAvroField(field));
+    }
+    return builder.build();
+  }
+
+  /**
+   * Given an Avro {@link GenericRecord} object, returns a protocol-buffer message that can be used
+   * to write data using the BigQuery Storage streaming API.
+   *
+   * @param descriptor The Descriptor for the DynamicMessage result
+   * @param record An Avro GenericRecord
+   * @return A dynamic message representation of a Proto payload to be used for StorageWrite API
+   */
+  public static DynamicMessage messageFromGenericRecord(
+      Descriptor descriptor, GenericRecord record) {
+    Schema schema = record.getSchema();
+    DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+    for (Schema.Field field : schema.getFields()) {
+      FieldDescriptor fieldDescriptor =
+          Preconditions.checkNotNull(descriptor.findFieldByName(field.name().toLowerCase()));
+      @Nullable
+      Object value =
+          messageValueFromGenericRecordValue(fieldDescriptor, field, field.name(), record);
+      if (value != null) {
+        builder.setField(fieldDescriptor, value);
+      }
+    }
+    return builder.build();
+  }
+
+  private static TableFieldSchema fieldDescriptorFromAvroField(Schema.Field field) {
+    @Nullable Schema schema = field.schema();
+    TableFieldSchema.Builder builder =
+        TableFieldSchema.newBuilder().setName(field.name().toLowerCase());
+    Schema elementType = null;
+    switch (schema.getType()) {
+      case RECORD:
+        if (schema == null) {
```

Comment:
If this were null, the above switch would already have thrown. Did you mean to check something else?

Reply:
Good catch, moved to before the switch statement.